packages/ai/src/generate-object/stream-object.ts

```diff
@@ -499,13 +503,35 @@ class DefaultStreamObjectResult<PARTIAL, RESULT, ELEMENT_STREAM>
           messages,
         } as Prompt);
 
+        const jsonSchema = await outputStrategy.jsonSchema();
+        const useToolMode = mode === 'tool';
+        const toolName = schemaName ?? 'json';
+
         const callOptions = {
-          responseFormat: {
-            type: 'json' as const,
-            schema: await outputStrategy.jsonSchema(),
-            name: schemaName,
-            description: schemaDescription,
-          },
+          responseFormat:
+            useToolMode || jsonSchema == null
+              ? undefined
+              : {
+                  type: 'json' as const,
+                  schema: jsonSchema,
+                  name: schemaName,
+                  description: schemaDescription,
+                },
+          tools:
+            useToolMode && jsonSchema != null
+              ? [
+                  {
+                    type: 'function' as const,
+                    name: toolName,
+                    description:
+                      schemaDescription ?? 'Respond with a JSON object.',
+                    inputSchema: jsonSchema,
+                  },
+                ]
+              : undefined,
```

Comment:
I'm not sure if we can inject a tool call like this. I think it's something that the AI SDK did in v4 and we moved away from it, please give me some time to investigate what happened